### PR TITLE
Mov/extensioncommunication

### PR DESCRIPTION
**Pull request by movccd**
## Purpose

_Describe the purpose of the pull request, link to issue describing the problem, etc.

## Changes

_Describe the changes that were needed.

## Checklist

- [ ] My code follows the style of this project.
- [ ] The code compiles without warnings.
- [ ] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

## CLA acceptance

_Remove if not applicable.

By submitting the contribution I accept the terms and conditions of the
Contributor License Agreement v1.0
- link: https://developers.concordium.com/CLAs/Contributor-License-Agreement-v1.0.pdf

- [ ] I accept the above linked CLA.

### CODE REVIEW

**Review by soerenbf — CHANGES_REQUESTED** [review]
Here is some stuff to look into:

- [ ] Remove prettierconfig from `.eslintrc.js`
- [ ] Use yarn to manage workspace dependencies, so we all use the same yarn.lock file
- [ ] Make sure there are no lint errors in the project. Try running `npx eslint_d . --ext .ts,.tsx --max-warnings 0` from the repository root.

Problems I ran into:

- Content script is not executing. Getting an error in the "Errors" page for the extension inside google chrome. As such, it's not possible to test the functionality, as nothing is injected onto the page. Maybe we can look into this together?
  - For some reason it works now... unpacked extensions seem to be a bit unstable.. getting weird errors once in a while that require restarting chrome or some other workaround.

File: packages/browser-wallet-message-hub/src/message-handlers/abstract-messagehandler.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    // Only in play if inheritor wants to be a conversation starter vs being a "listener,responder".
+    private publisherPort$?: chrome.runtime.Port;
+
+    protected constructor(me: HandlerTypeEnum) {
+        super();
+        this.me$ = me;
+    }
+
+    // Event listener for port messages
+    public async onPortMessage(message: Message, port: chrome.runtime.Port): Promise<void> {
```
Comment (soerenbf) [inline packages/browser-wallet-message-hub/src/message-handlers/abstract-messagehandler.ts:25]: Why is this public? Will I ever need this as a consumer of the API?

File: packages/browser-wallet-message-hub/src/message-handlers/message.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+export class Message {
+    public correlationId: string;
+
+    public readonly ccFilterMarker: string = filterMarkerGuid;
+
+    constructor(
+        public from: HandlerTypeEnum,
+        public to: HandlerTypeEnum,
+        public messageType: MessageTypeEnum,
+        public payload: any
```
Comment (soerenbf) [inline packages/browser-wallet-message-hub/src/message-handlers/message.ts:16]: Should be optional, as messages without payloads can be valid (f.x. init)

```suggestion
        public payload?: any
```

**Review by soerenbf — COMMENTED** [review]
Couldn't add to my existing review, so here comes a 2nd one.

Some things I find weird from trying to use the message-hub:

- When publishing messages, I have to construct messages, declaring who the message is from. Why is this not given from the message handler used (f.x. all messages from the background handler will have to = backgroundScript)? would be easier to use the handlers if they contained factories for building a message for that particular script.
- There's a lot of setup outside of constructing the message handler. Why is this not just part of the constructor? Makes it hard to 
  - `contentMesageHandler.addWindowPostMessageEventListemer()`
  - `contentMesageHandler.createPortAndSetupEventListeners()`
- Would be easier to subscribe to messages, if it didn't go through the `EventEmitter` interface, as I need to specify an arbitrary event type (string), which can be anything I guess?

File: packages/browser-wallet-message-hub/src/message-handlers/abstract-wallet-messagehandler.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        port.onMessage.removeListener(this.onPortMessage.bind(this));
+
+        if (port.sender?.tab?.id) {
+            this.tabsDictionary.delete(port.sender.tab.id);
+        }
+    }
+
+    // Public
+
+    public addRuntimePortListeners(): void {
```
[the remark this answers is not in the document]
Comment (movccd) [reply packages/browser-wallet-message-hub/src/message-handlers/abstract-wallet-messagehandler.ts:33]: I havent added any JSDoc yet as this is WIP. We would probably end up rewriting JSDoc several times before nailing it. But if doing JSDoc on prelim stuff is standard - I will certainly do it.


The operation is exposed to make it more explicit in the "client" code.

File: packages/browser-wallet-api/src/wallet-api.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import { EventEmitter } from 'eventemitter3';
+import { logger } from '@concordium/browser-wallet-message-hub/src/message-handlers/logger';
+
+export interface IWalletApi {
+    sendTransaction(): Promise<Message>;
+    signMessage(): Promise<Message>;
+    getAccounts(): Promise<Message>;
+}
+
+class WalletApi extends EventEmitter implements IWalletApi {
```
[the remark this answers is not in the document]
Comment (movccd) [reply packages/browser-wallet-api/src/wallet-api.ts:17]: Consumers are not intended to create instance. There should be only one instance which should be assigned to window.concordium

File: packages/browser-wallet-message-hub/src/message-handlers/messagetype-enum.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+ */
+export enum MessageTypeEnum {
+    // Methods
+    init = 'init',
+    sendTransaction = 'sendTransaction',
+    signMessage = 'SignMessage',
+    getAccounts = 'GetAccounts',
+
+    // Events
+    event = 'EventMessage',
```
[the remark this answers is not in the document]
Comment (movccd) [reply packages/browser-wallet-message-hub/src/message-handlers/messagetype-enum.ts:12]: will be fixed to PascalCase - all the way through

File: packages/browser-wallet-api/src/wallet-api.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+export const walletApi = new WalletApi(new InjectedMessageHandler());
+
+// TODO: Just remove. Used for testing only
+setInterval(async () => {
+    await walletApi.sendTransaction().then((m) => {
+        logger.log(JSON.stringify(m));
+    });
+}, 5000);
```
[the remark this answers is not in the document]
Comment (movccd) [reply packages/browser-wallet-api/src/wallet-api.ts:80]: agree

File: packages/browser-wallet-message-hub/src/message-handlers/abstract-wallet-messagehandler.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (soerenbf) [reply packages/browser-wallet-message-hub/src/message-handlers/abstract-wallet-messagehandler.ts:33]: I think the simpler we can make the public API of these message handlers, the better. Ideally, I'd want to:

- Create an instance
- Subscribe to messages + send response to message
- Remove subscription(s)
- Publish message

In my world, only these 4 things should pop up as things to do when i try to access the public API.

File: packages/browser-wallet-message-hub/src/message-handlers/messagetype-enum.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (soerenbf) [reply packages/browser-wallet-message-hub/src/message-handlers/messagetype-enum.ts:12]: The `HandlerEnum` is also camelCased, so remember to change that as well then :wink:

**Review by soerenbf — APPROVED** [review]

**Pull request merged**